File: Android_kotlin_app_MVVM/app/src/test/java/com/mindsilence/game/domain/repository/InMemoryGameProgressRepository.kt
package com.mindsilence.game.domain.repository // In-memory fake; production code never imports this.

import com.mindsilence.game.domain.model.DailyStats // Same row type as the prefs implementation.
import java.time.LocalDate // Test clock is the JVM default.

/** In-memory [GameProgressRepository] for unit tests; nothing is written to disk. */
class InMemoryGameProgressRepository : GameProgressRepository { // Used by ViewModel and UseCase tests.

    private val statsByDate = linkedMapOf<LocalDate, DailyStats>() // Insertion order; sorted on read.

    override fun recordSession(levelReached: Int, totalSeconds: Int): Int { // Same merge rules as prefs.
        val today = LocalDate.now() // Test clock is the JVM default.
        val current = statsByDate[today] // Null if first today.
        val updated = if (current == null) { // First attempt today.
            DailyStats( // New row.
                date = today, // Day key.
                attempts = 1, // This attempt.
                totalSeconds = totalSeconds, // This length.
                bestLevel = levelReached, // This level.
            )
        } else { // Same day again.
            current.copy( // Merge.
                attempts = current.attempts + 1, // Count.
                totalSeconds = current.totalSeconds + totalSeconds, // Sum.
                bestLevel = maxOf(current.bestLevel, levelReached), // Max level.
            )
        }
        statsByDate[today] = updated // Replace today’s row.
        return updated.bestLevel // Same return as the prefs impl.
    }

    override fun getDailyStats(): List<DailyStats> = // Newest first, like prefs.
        statsByDate.values.sortedByDescending { it.date } // Highscore order.
}

File: Unity_quest_app/Assets/Scripts/Data/InMemoryGameProgressRepository.cs
using System;
using System.Collections.Generic;
using MindSilence.Domain;

namespace MindSilence.Data
{
    public sealed class InMemoryGameProgressRepository : IGameProgressRepository
    {
        private readonly Func<DateTime> _today;
        private readonly Dictionary<DateTime, DailyStats> _statsByDate = new Dictionary<DateTime, DailyStats>();

        public InMemoryGameProgressRepository(Func<DateTime> today = null)
        {
            _today = today ?? (() => DateTime.Now.Date);
        }

        public int RecordSession(int levelReached, int totalSeconds)
        {
            var today = _today().Date;
            DailyStats updated;
            if (!_statsByDate.TryGetValue(today, out var current))
            {
                updated = new DailyStats(today, attempts: 1, totalSeconds, levelReached);
            }
            else
            {
                var bestLevel = current.BestLevel > levelReached ? current.BestLevel : levelReached;
                updated = new DailyStats(
                    today,
                    current.Attempts + 1,
                    current.TotalSeconds + totalSeconds,
                    bestLevel);
            }

            _statsByDate[today] = updated;
            return updated.BestLevel;
        }

        public IReadOnlyList<DailyStats> GetDailyStats()
        {
            var stats = new List<DailyStats>(_statsByDate.Values);
            stats.Sort((left, right) => right.Date.CompareTo(left.Date));
            return stats;
        }
    }
}
